fixed-hash/src/hash.rs

```diff
@@ -16,157 +20,224 @@ pub fn clean_0x(s: &str) -> &str {
 }
 
 /// Construct a fixed-size hash type.
-/// Takes the name of the type and the size in bytes and an optional third argument for meta data
-/// Example: `construct_hash!(H256, 32);`
-/// Example: `construct_hash!(H160, 20, cfg_attr(feature = "serialize", derive(Serialize, Deserialize)));`
+///
+/// # Examples
+///
+/// Create a public unformatted hash type with 32 bytes size.
+///
+/// ```
+/// # #[macro_use] extern crate fixed_hash;
+/// construct_hash!{ pub struct H256(32); }
+/// # fn main() {
+/// # 	assert_eq!(::std::mem::size_of::<H256>(), 32);
+/// # }
+/// ```
+///
+/// With additional attributes and doc comments.
+///
+/// ```
+/// # #[macro_use] extern crate fixed_hash;
+/// construct_hash!{
+/// 	/// My unformatted 160 bytes sized hash type.
+/// 	#[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
+/// 	pub struct H160(20);
+/// }
+/// # fn main() {
+/// # 	assert_eq!(::std::mem::size_of::<H160>(), 20);
+/// # }
+/// ```
+///
+/// As always, visibility modifier is optional.
+/// So for a private type go ahead with this.
+///
+/// ```
+/// # #[macro_use] extern crate fixed_hash;
+/// construct_hash!{ struct H512(64); }
+/// # fn main() {
+/// # 	assert_eq!(::std::mem::size_of::<H512>(), 64);
+/// # }
+/// ```
 #[macro_export]
 macro_rules! construct_hash {
-	($from: ident, $size: expr $(, $m:meta)*) => {
+	( $(#[$attr:meta])* $visibility:vis struct $name:ident ( $n_bytes:expr ); ) => {
 		#[repr(C)]
-		$(#[$m])*
-		/// Unformatted binary data of fixed length.
-		pub struct $from (pub [u8; $size]);
+		$(#[$attr])*
+		$visibility struct $name (pub [u8; $n_bytes]);
 
-
-		impl From<[u8; $size]> for $from {
-			fn from(bytes: [u8; $size]) -> Self {
-				$from(bytes)
-			}
-		}
-
-		impl From<$from> for [u8; $size] {
-			fn from(s: $from) -> Self {
-				s.0
+		impl From<[u8; $n_bytes]> for $name {
+			#[inline]
+			fn from(bytes: [u8; $n_bytes]) -> Self {
+				$name(bytes)
 			}
 		}
 
-		impl ::core::ops::Deref for $from {
-			type Target = [u8];
-
+		impl From<$name> for [u8; $n_bytes] {
 			#[inline]
-			fn deref(&self) -> &[u8] {
```

Comment:
According to the Rust documentation and API style guidelines `Deref` and `DerefMut` implementations are meant exclusively for smart pointer types.

> Implementing Deref for smart pointers makes accessing the data behind them convenient, which is why they implement Deref. On the other hand, the rules regarding Deref and DerefMut were designed specifically to accommodate smart pointers. Because of this, Deref should only be implemented for smart pointers to avoid confusion.

Reply:
In version 0.3 we still have `as_ref` and the new `as_bytes` methods to replace all uses of `Deref` and `DerefMut`. :)

Reply:
makes sense

Reply:
resolved then